{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import packages\n",
    "#---------------------------------------\n",
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import warnings\n",
    "import random\n",
    "import arviz as az\n",
    "import pymc as pm\n",
    "warnings.filterwarnings(\"ignore\", category=RuntimeWarning) \n",
    "\n",
    "#Import your modules\n",
    "#---------------------------------------\n",
    "import admin_functions as adfn\n",
    "import cell_decomp_func as cdfn\n",
    "\n",
    "# Define paths\n",
    "#----------------------------------------------------------------------\n",
    "l_code = '/Users/dominicburrows/Dropbox/PhD/Analysis/my_scripts/GitHub/'\n",
    "l_data = '/Users/dominicburrows/Dropbox/PhD/analysis/Project/'\n",
    "l_fig = '/Users/dominicburrows/Dropbox/PhD/figures/'\n",
    "\n",
    "s_code = '/sphere/dburrows12/'\n",
    "s_data = '/sphere/dburrows12/'\n",
    "\n",
    "%load_ext autoreload\n",
    "sys.version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Noise model - Linear regression with noise covariates\n",
    "#standard params\n",
    "#n_clusts = 10\n",
    "#n_genes = 600\n",
    "#n_cells = 500\n",
    "#rate_range = 0,20\n",
    "#dropout = 85%\n",
    "#E_std = 9\n",
    "#a_std = 6\n",
    "#g_std = 8\n",
    "\n",
    "#Define parameters of simulated data\n",
    "n_clusts = 10\n",
    "n_genes = 600\n",
    "n_cells = 500\n",
    "rate_range = 0,20 #max and min of uniform distribution for generating rates\n",
    "per = 85 #percentage of dropped genes\n",
    "e_std= 9 #spot + gene noise\n",
    "g_std = 7 #gene specific noise\n",
    "a_std = 6 #spot specific noise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Simulate spot data from simulated gene expression\n",
    "spot_sim = cdfn.simulate_cell_mix(n_clusts, n_cells, n_genes).simulate_gene_exp(rate_range)\n",
    "n_spots = spot_sim.__dict__['n_spots']\n",
    "spots = spot_sim.__dict__['spots']\n",
    "ref_exp = spot_sim.__dict__['mean_exps']\n",
    "prop_vec = spot_sim.__dict__['prop_vec']\n",
    "\n",
    "#Dropout a certain percentage of genes\n",
    "for i in range(spots.shape[0]):\n",
    "    rand_ind = np.random.choice(np.arange(spots.shape[1]), size = int((per/100) * spots.shape[1]), replace=False) #random index for selecting\n",
    "    spots[i,rand_ind]=0\n",
    "\n",
    "#Add random noise and make int and remove negatives\n",
    "spots = spots+np.random.normal(0, e_std, (spots.shape))\n",
    "\n",
    "#gamma - over each gene\n",
    "gamma = np.random.normal(0, g_std, (spots.shape[1]))\n",
    "gamma_mat = np.asarray([gamma for i in range(spots.shape[0])]) #Repeat across columns for elementwise addition\n",
    "spots = spots+gamma_mat\n",
    "\n",
    "#alpha - over each spot\n",
    "alpha = np.random.normal(0, a_std, (spots.shape[0]))\n",
    "alpha_mat = np.asarray([alpha for i in range(spots.shape[1])]).T #Repeat across columns for elementwise addition\n",
    "spots = spots+alpha_mat\n",
    "\n",
    "spots = spots.astype(int) \n",
    "spots[spots < 0] = 0\n",
    "spots +=1 #remove any zeros\n",
    "\n",
    "\n",
    "#Poisson GLM with noise\n",
    "Poisson_noise_GLM=pm.Model()\n",
    "\n",
    "with Poisson_noise_GLM:\n",
    "\n",
    "    beta=pm.HalfNormal(\"beta\", sigma=1, shape=(n_spots, n_clusts)) # proportions of each cell at each spot\n",
    "    eps=pm.Normal(\"eps\", mu= 0, sigma=1, shape=(spots.shape)) # random noise at each spot and gene\n",
    "    gamma = pm.Normal(\"gamma\", mu= 0, sigma=1, shape=(spots.shape[1])) # random noise at each gene\n",
    "    alpha = pm.Normal(\"alpha\", mu= 0, sigma=1, shape=(spots.shape[0])) # random noise at each spot\n",
    "    \n",
    "    lmd=pm.math.dot(beta, ref_exp) #each spot is a linear combination of cell proportions and their reference gene expression\n",
    "    lmd = lmd + eps + np.asarray([gamma.eval() for i in range(spots.shape[0])]) + np.asarray([alpha.eval() for i in range(spots.shape[1])]).T #Add noise\n",
    "\n",
    "    #Likelihood of observed data given Poisson rates\n",
    "    y=pm.Poisson(\"y\", mu=lmd, observed=spots)\n",
    "\n",
    "with Poisson_noise_GLM:\n",
    "    idata=pm.sample(draws=1000, chains=1)\n",
    "\n",
    "mean_post = np.mean(idata.posterior['beta'][0],axis=0)\n",
    "Nd = np.sum(mean_post, axis=1) \n",
    "Nd = np.asarray([Nd for i in range(n_clusts)]).T \n",
    "mean_post = np.divide(np.mean(idata.posterior['beta'][0],axis=0),Nd)\n",
    "from scipy.stats import linregress\n",
    "line_fit=linregress(np.ravel(prop_vec), np.ravel(mean_post))\n",
    "r2 = line_fit.rvalue\n",
    "\n",
    "savename = 'RCTD-test_PARSTAND_ALL.npy'\n",
    "np.save(s_data + 'SPATIAL-TRANSCRIPTOMICS/' + savename, [idata, mean_post, prop_vec, r2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.16 (main, Jan 11 2023, 16:05:54) \n[GCC 11.2.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "980ae2358fca277a1e6b7c0afec83352662af6bf9b787ebbb80b8db147d70624"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
